### PR TITLE
all: migrate gradle build to java-library plugin

### PR DESCRIPTION
**Pull request by dapengzhang0**
- Use gradle configuration `api` for dependencies that are part of grpc public api signatures.
- Replace deprecated gradle configurations `compile`, `testCompile`, `runtime` and `testRuntime`.
- Exclude guava's transitive dependencies jsr305 and animal-sniffer-annotations, and always manually add them as runtimeOnly dependency. `error_prone_annotations` is an exception: It is also excluded but manually added not as runtimeOnly. It must always compile with guava, otherwise users will see warning spams if guava is in the compile classpath but error_prone_annotations is not (See https://github.com/google/guava/issues/2721).

### CODE REVIEW

**Comment by dapengzhang0** [comment]
@ejona86 Now is ready for review. PTAL.

File: alts/build.gradle [hunk 1]
```diff
@@ -13,35 +13,43 @@ sourceCompatibility = 1.7
...
     }
-    compileOnly libraries.javax_annotation
-
-    shadow configurations.compile.getDependencies().minus(nettyDependency)
-    shadow project(path: ':grpc-netty-shaded', configuration: 'shadow')
-
-    testCompile project(':grpc-testing'),
+    implementation (libraries.guava) {
+        // prefer our own version from grpc-api instead of Guava's dependency
+        exclude group: 'com.google.code.findbugs', module: 'jsr305'
```
Comment (ejona86) [inline alts/build.gradle:33]: As we talked about earlier, I really want to see excluded dependencies listed as an explicit dependency.

Comment (dapengzhang0) [reply alts/build.gradle]: Done with a util function
https://github.com/grpc/grpc-java/pull/6902/files#diff-c197962302397baf3a4cc36463dce5eaR191-R203

Exclude guava's transitive dependencies jsr305 and animal-sniffer-annotations, and always manually add them as runtimeOnly dependency. error_prone_annotations is an exception: It is also excluded but manually added not as runtimeOnly. It must alway compile with guava, otherwise users will see warning spams if guava is in the compile classpath but error_prone_annotations is not (See google/guava#2721).

Comment (ejona86) [reply alts/build.gradle]: That's a nice solution.

File: alts/build.gradle [hunk 2]
```diff
@@ -13,35 +13,35 @@ sourceCompatibility = 1.7
...
+    testImplementation project(':grpc-testing'),
             project(':grpc-testing-proto'),
             libraries.guava,
             libraries.guava_testlib,
             libraries.junit,
             libraries.mockito,
             libraries.truth
-    testRuntime libraries.netty_tcnative,
-            libraries.netty_epoll
+    testRuntimeOnly libraries.netty_epoll,
```
Comment (ejona86) [inline alts/build.gradle:39]: Note: unnecessary changes like reordering deps and moving the type of deps (like `shadow` here) around made reviewing this PR significantly harder. Sorting dependencies is nice, but would have been much better as a separate PR.

File: gae-interop-testing/gae-jdk8/build.gradle [hunk 3]
```diff
@@ -43,17 +43,15 @@ repositories {
...
-    compile (project(":grpc-interop-testing")) {
-        exclude group: "io.grpc", module: "grpc-netty-shaded"
+    implementation project(':grpc-netty'),
+            project(":grpc-stub"),
+            libraries.protobuf,
+            libraries.truth
+    implementation (project(':grpc-interop-testing')) {
+        exclude group: 'io.grpc', module: 'grpc-alts'
     }
-    compile libraries.netty_tcnative
```
Comment (ejona86) [inline gae-interop-testing/gae-jdk8/build.gradle:56]: This looks like it was removed. It seems it should be a runtimeOnly dep.

File: grpclb/build.gradle [hunk 4]
```diff
@@ -11,17 +11,17 @@ description = "gRPC: GRPCLB LoadBalancer plugin"
 evaluationDependsOn(project(':grpc-core').path)
 
 dependencies {
-    compile project(':grpc-core'),
+    compileOnly libraries.javax_annotation
+    implementation project(':grpc-core'),
             project(':grpc-protobuf'),
             project(':grpc-stub'),
```
Comment (ejona86) [inline grpclb/build.gradle:17]: I'm surprised this trailing comma doesn't cause parsing problems. (Travis isn't failing, so it seems it really isn't a problem, but it makes me wonder.)

File: grpclb/build.gradle [hunk 5]
```diff
@@ -11,17 +11,17 @@ description = "gRPC: GRPCLB LoadBalancer plugin"
...
-            libraries.protobuf
-    compile (libraries.protobuf_util) {
-        // prefer our own versions instead of protobuf-util's dependency
-        exclude group: 'com.google.guava', module: 'guava'
+    implementation (libraries.protobuf_util) {
+        // prefer our own version from grpc-stub instead of protobuf-util's dependency
         exclude group: 'com.google.errorprone', module: 'error_prone_annotations'
```
Comment (ejona86) [inline grpclb/build.gradle:20]: We need to add explicitly dependencies on errorprone and guava to match these excluded ones.

File: grpclb/build.gradle [hunk 6]
```diff
@@ -11,17 +11,17 @@ description = "gRPC: GRPCLB LoadBalancer plugin"
 evaluationDependsOn(project(':grpc-core').path)
 
 dependencies {
-    compile project(':grpc-core'),
+    compileOnly libraries.javax_annotation
+    implementation project(':grpc-core'),
             project(':grpc-protobuf'),
             project(':grpc-stub'),
-            libraries.protobuf
```
Comment (voidzcy) [inline grpclb/build.gradle:17]: Why removed direct dependency of protobuf?

File: build.gradle [hunk 7]
```diff
@@ -185,18 +187,76 @@ subprojects {
             // Jetty ALPN dependencies
             jetty_alpn_agent: 'org.mortbay.jetty.alpn:jetty-alpn-agent:2.0.9'
         ]
+
+        // A util function to config guava dependency with the given configuration
```
[the remark this answers is not in the document]
Comment (dapengzhang0) [reply build.gradle:191]: Having each subproject list transitive dependency manipulation is easy to break by mistake as code evolves. For example, in one subproject there could be multiple dependencies excluding the same transitive dependencies. You either have to duplicate lines to add excluded items or scatter them in separate places.
```groovy
    implementation(libraries.guava) {
        // prefer our own versions from libraries instead of Guava's dependency
        exclude group: 'com.google.errorprone', module: 'error_prone_annotations'
        exclude group: 'com.google.code.findbugs', module: 'jsr305'
        exclude group: 'org.codehaus.mojo', module: 'animal-sniffer-annotations'
    }
    runtimeOnly libraries.jsr305
    implementation libraries.errorprone
    runtimeOnly libraries.animal-sniffer

    implementation (libraries.opencensus_api) {
        // prefer 3.0.2 from libraries instead of 3.0.1
        exclude group: 'com.google.code.findbugs', module: 'jsr305'
        // prefer 20.0 from libraries instead of 19.0
        exclude group: 'com.google.guava', module: 'guava'
        // we'll always be more up-to-date
        exclude group: 'io.grpc', module: 'grpc-context'
    }
    // duplicate line, otherwise you may forgot when other dependencies are removed.
    runtimeOnly libraries.jsr305
    // duplicate lines, otherwise you may forgot when other dependencies are removed.
    runtimeOnly(libraries.guava) {
        // prefer our own versions from libraries instead of Guava's dependency
        exclude group: 'com.google.errorprone', module: 'error_prone_annotations'
        exclude group: 'com.google.code.findbugs', module: 'jsr305'
        exclude group: 'org.codehaus.mojo', module: 'animal-sniffer-annotations'
    }
    // duplicate line, otherwise you may forgot when those in other dependencie are removed.
    runtimeOnly libraries.jsr305
    runtimeOnly libraries.errorprone
    runtimeOnly libraries.animal-sniffer
    runtimeOnly project(':grpc-context')

    implementation (libraries.opencensus_contrib_grpc_metrics) {
        // prefer 3.0.2 from libraries instead of 3.0.1
        exclude group: 'com.google.code.findbugs', module: 'jsr305'
        // we'll always be more up-to-date
        exclude group: 'io.grpc', module: 'grpc-context'
        // prefer 20.0 from libraries instead of 19.0
        exclude group: 'com.google.guava', module: 'guava'
    }
    // duplicate line, otherwise you may forgot when other dependencies are removed.
    runtimeOnly libraries.jsr305
    // duplicate lines, otherwise you may forgot when other dependencies are removed.
    runtimeOnly(libraries.guava) {
        // prefer our own versions from libraries instead of Guava's dependency
        exclude group: 'com.google.errorprone', module: 'error_prone_annotations'
        exclude group: 'com.google.code.findbugs', module: 'jsr305'
        exclude group: 'org.codehaus.mojo', module: 'animal-sniffer-annotations'
    }
    // duplicate line, otherwise you may forgot when those in other place are removed.
    runtimeOnly libraries.jsr305
    runtimeOnly libraries.errorprone
    runtimeOnly libraries.animal-sniffer
    runtimeOnly project(':grpc-context')
```

[the remark this answers is not in the document]
Comment (voidzcy) [reply build.gradle:191]: I don't understand the example you posted (why Guava is added to implementation/runtimeOnly repeatedly?). But my point was what we currently have (not using utility functions) looks pretty straightforward as each subproject's buildscript clearly shows its dependencies, without redirecting to somewhere else to see what is included/excluded. Just kind of styling preference.

[the remark this answers is not in the document]
Comment (dapengzhang0) [reply build.gradle:191]: >why Guava is added to implementation/runtimeOnly repeatedly?

The first Guava is added because the subproject needs it. The second Guava added is because it is excluded from libraries.opencensus_api you need added it back. If you don't added immediately after exclusion because it is already added somewhere els, then it is bug-prone as the build.gradle evolves.

[the remark this answers is not in the document]
Comment (ejona86) [reply build.gradle:191]: The `guava` dependency is an internal implementation detail, so it goes in the list of `implementation` deps. But that means it will not be automatically inherited to down-stream projects (which is what we want), including our own ("Oh.").

> having each subproject clearly list out its transitive dependency manipulation still seems to be more elegant, given that most users use a subset of subprojects.

`guavaDependency 'implementation'` is semantically equivalent to `implementation libraries.guava`. The difference is that the transitive dependencies versions don't match and `resolutionStrategy.failOnVersionConflict()` will fail the build. We have that check to avoid issues where Maven would choose the wrong version (since Maven chooses the first version encountered, not the greatest version number).

To "fix" `failOnVersionConflict` we have to exclude dependencies. But that is dangerous since we are excluding dependencies based on what's used _anywhere_ in gRPC, so a user may not get the dependency _at all_ if it is excluded. To make it reliable, we add the excluded dependencies explicitly. That makes it _much_ easier to audit, and removes all risk of things getting out-of-sync cross-module.

Note that `runtimeOnly libraries.jsr305` doesn't provide any "value" to a module, since it can't be used for compilation. So this utility is actually a pretty fair abstraction. The only place it breaks down is for error-prone with Guava, because the annotations must be available during compilation and so it is an `implementation` dependency.

File: alts/build.gradle (hunk 2 above)
Comment (dapengzhang0) [reply alts/build.gradle:39]: Sorry for the noise and extra time and effort spent for the review because of it. I'll send smaller PRs without unnecessary changes like reordering deps.

File: gae-interop-testing/gae-jdk8/build.gradle (hunk 3 above)
Comment (ejona86) [reply gae-interop-testing/gae-jdk8/build.gradle:56]: Yeah, we use tcnative on GAE Java 8 and want to make sure it works, since most people will be using grpc-netty-shaded which includes it. Since v1.25.0 the test probably passes without tcnative, though, since GAE Java 8 includes Conscrypt. But we still want to test with tcnative.

Comment (dapengzhang0) [reply gae-interop-testing/gae-jdk8/build.gradle:56]: I see. Add it add as runtimeOnly dep in #7001 now.

**Comment by dapengzhang0** [comment]
Closing in favor of #7001.
